README.md
# pipapp
PipApp creates basic Python application directories and files for pip-distribution.


CHANGES
Version 0.4 (13.06.2016)

- Added find_packages() to generated setup.py.

Version 0.3 (28.01.2016)

- Added commented 'install_requires'-keyword to setup.py that gets generated.

Version 0.2 (27.01.2016)

- Improved command line parsing and usage string.

Version 0.1 (26.01.2016)

- Initial version.



pipapp/texts.py
# -*- coding: utf-8 -*-

# The MIT License (MIT)

# Copyright (c) 2016 Sami Salkosuo

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

import datetime


def getLicenseTxt(author):
    license="""The MIT License (MIT)

Copyright (c) %s %s

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.""" % (datetime.date.today().strftime("%Y"),author)
    
    return license


def getReadmeTxt(projectName="PROJECTNAME"):
    readme="""%s
======

Project description

Requirements
------------

Python 3.

Install
-------

Install latest version: **pip install %s**.

Usage
-----

Execute *%s*.


About
-----

""" % (projectName,projectName,projectName)

    return readme


def getSetupCfgTxt():
    text=["[bdist_wheel]","universal = 0"]
    return text


def getRunnerTxt(projectName):
    text='''#!/usr/bin/env python3
# -*- coding: utf-8 -*-


"""Convenience wrapper for running %s directly from source tree."""


from %s.%s import main


if __name__ == '__main__':
    main()
''' % (projectName,projectName,projectName)

    return text


def getManifestIn(projectname):
    txt='''# Include the license file
include LICENSE 
'''
    return txt

def getSetupPy(projectName,config):

    
    txt='''# -*- coding: utf-8 -*-
 
 
"""setup.py: setuptools control."""
 
 
import re
from setuptools import setup, find_packages

projectName="{pName:s}"
scriptFile="%s/%s.py" % (projectName,projectName)
description="Setuptools setup.py for {pName:s}."


version = re.search(
    '^__version__\s*=\s*"(.*)"',
    open(scriptFile).read(),
    re.M
    ).group(1)
 
 
with open("README.rst", "rb") as f:
    long_descr = f.read().decode("utf-8")
 
 
setup(
    name = projectName,
    packages = find_packages(),
    #add required packages to install_requires list
    #install_requires=["package","package2"]
    entry_points = {{
        "console_scripts": ['%s = %s.%s:main' % (projectName,projectName,projectName)]
        }},
    version = version,
    description = description,
    long_description = long_descr,
    author = "{author:s}",
    author_email = "{author_email:s}",
    url = "{url:s}",
    license='{license:s}',
#list of classifiers: https://pypi.python.org/pypi?%3Aaction=list_classifiers
    classifiers={classifiers:s},
    )
'''.format(pName=projectName,author=config['author'],author_email=config['author_email'],url=config['url'],license=config['license'],classifiers=str(config['classifiers']).replace(",",",\n"))

    return txt


def getInitPy(projectName):
    #empty __init__.py is enough
    txt=""
    return txt  


def getMainPy(projectName):
    txt='''# -*- coding: utf-8 -*-
 
"""{pName:s}.__main__: executed when {pName:s} directory is called as script."""
 
from .{pName:s} import main
main()

'''.format(pName=projectName)


    return txt  

def getMainSource(projectName,license,config):
    txt='''#!/usr/bin/env python3
# -*- coding: utf-8 -*-


'''
    for line in license.split("\n"):
        txt="%s# %s\n" % (txt,line)



    txt=txt+'''
#add correct version number here
__version__ = "0.0.1"


PROGRAMNAME="{pName:s}"
VERSION=__version__
COPYRIGHT="(C) {currentYear:s} {author:s}"


def main():
    #program logic here
    print("TODO: the application")
    pass

if __name__ == "__main__": 
    main()

'''.format(pName=projectName,currentYear=datetime.date.today().strftime("%Y"),author=config['author'])




    return txt

pipapp/pipapp.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# The MIT License (MIT)

# Copyright (c) 2016 Sami Salkosuo

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

#add correct version number here
__version__ = "0.4"

PROGRAMNAME="PipApp"
VERSION=__version__
COPYRIGHT="(C) 2016 Sami Salkosuo"
LICENSE="Licensed under The MIT License."
DESCRIPTION="Create files and directories for pip-dist enabled Python apps."
CONFIG_FILE=".pipapp_defaults"

import argparse
import configparser
import os
import datetime
import sys
from .texts import *

from os.path import expanduser

BASEDIR="."
PROJECTNAME=None
#command line args
args=None

def parseCommandLineArgs():
    #parse command line args
    parser = argparse.ArgumentParser(description='PipApp. %s' % DESCRIPTION)
    parser.add_argument('-d','--dir', nargs=1,metavar='DIR', help='Root directory where to create new project files and dirs. Default is current directory.')
    parser.add_argument('-v,--version', action='version', version="%s v%s" % (PROGRAMNAME, VERSION))
    parser.add_argument("projectname",metavar='PROJECTNAME', nargs=1)
    global args
    args = parser.parse_args()

def readConfig():
    homeDir = expanduser("~")
    configFile="%s/%s" % (homeDir,CONFIG_FILE)
    config = configparser.ConfigParser()

    if os.path.isfile(configFile) == False:
        #create default configrutaion
        cfg={}
        cfg['author']="Your Name"
        cfg['author_email']="Your Email"
        cfg['url']="http://Project.url.here"
        cfg['license']="MIT"
        cfg['classifiers']=["Development Status :: 1 - Planning",
                            "License :: OSI Approved :: MIT License",
                            "Environment :: Console",
                            "Natural Language :: English",
                            "Operating System :: OS Independent",
                            "Programming Language :: Python :: 3.4",
                            "Programming Language :: Python :: 3.5",
                            "Programming Language :: Python :: 3 :: Only"]


        config['DEFAULT'] = cfg
        with open(configFile, 'w') as configfile:
            config.write(configfile)

    #read configurato
    config.read(configFile)

    return config["DEFAULT"]


def createFile(fileName,fileContents,directory):
    print("Generating %s/%s...." % (directory,fileName), end="")
    f=open("%s/%s" % (directory,fileName) ,"w")
    if isinstance(fileContents, (list)):
        for line in fileContents:
            print(line,file=f)
    else:
        print(fileContents,file=f,end="")
    f.close()
    print("Done.")

def createDirectory(directoryName):
    if not os.path.exists(directoryName):
        os.makedirs(directoryName)

def main():

    parseCommandLineArgs()

    PROJECTNAME=args.projectname[0]
    global BASEDIR
    BASEDIR=PROJECTNAME
    if args.dir:
        BASEDIR="%s/%s" % (args.dir[0],PROJECTNAME)
    createDirectory(BASEDIR)

    print("Generating files for project %s in dir %s..." % (PROJECTNAME,BASEDIR))

    config=readConfig()

    #file: README.rst
    createFile("README.rst",getReadmeTxt(PROJECTNAME),BASEDIR)

    #file: CHANGES
    currentDate=datetime.date.today().strftime("%d.%m.%Y")
    changes=["Version 0.1 (%s)" % currentDate,"","- Initial version."]
    createFile("CHANGES",changes,BASEDIR)

    #file: LICENSE
    license=getLicenseTxt(config['author'])
    createFile("LICENSE",license,BASEDIR)
    
    #file: setup.cfg
    createFile("setup.cfg",getSetupCfgTxt(),BASEDIR)

    #file: PROJECTNAME-runner.py
    createFile("%s-runner.py" % PROJECTNAME,getRunnerTxt(PROJECTNAME),BASEDIR)

    #file: setup.py
    createFile("setup.py",getSetupPy(PROJECTNAME,config),BASEDIR)

    #file: MANIFEST.in
    createFile("MANIFEST.in",getManifestIn(PROJECTNAME),BASEDIR)

    #create dir: PROJECTNAME
    srcDir=BASEDIR+"/"+PROJECTNAME
    print("Creating dir: %s..." % srcDir, end="")
    createDirectory(srcDir)
    print("Done.")

    #file: PROJECTNAME/PROJECTNAME.py
    createFile("%s.py" % PROJECTNAME,getMainSource(PROJECTNAME,license,config),srcDir)
    
    #file: PROJECTNAME/__init__.py
    createFile("__init__.py" ,getInitPy(PROJECTNAME),srcDir)

    #file: PROJECTNAME/__main__.py
    createFile("__main__.py" ,getMainPy(PROJECTNAME),srcDir)



if __name__ == "__main__": 
    main()
